File: maui/KsDialogs.Maui/Contract/DialogOutcome.cs
namespace KsDialogs;

/// <summary>
/// 結果の型消去表現。
/// </summary>
/// <remarks>
/// 公開 API の <see cref="DialogResult{TResult}"/> と違い結果値の型を持たない、
/// レジストリ・提示層・委譲面が共通に使う輸送形。宣言結果型への復元は <see cref="Dialog"/> が行う。
/// </remarks>
internal abstract record DialogOutcome
{
    // 派生を Completed / Cancelled の 2 つに閉じるため、基底のコンストラクタは入れ子の型からしか呼べない
    private DialogOutcome()
    {
    }

    /// <summary>結果値つきの完了。</summary>
    /// <param name="Value">報告された結果値。</param>
    public sealed record Completed(object? Value) : DialogOutcome;

    /// <summary>結果値を持たないキャンセル。</summary>
    public sealed record Cancelled : DialogOutcome
    {
        /// <summary>値を持たないため使い回せる唯一の実体。</summary>
        public static Cancelled Instance { get; } = new();
    }
}

File: android/ksdialogs/src/main/kotlin/jp/kamusoft/ksdialogs/DialogOutcome.kt
package jp.kamusoft.ksdialogs

/**
 * 結果の型消去表現。
 *
 * 公開 API の [DialogResult] と違い結果値の型を持たず、レジストリ・提示層が共通に使う輸送形。
 * 値の実体は [DialogNotifier] が宣言結果型で受け取ったものをそのまま運ぶ。
 */
internal sealed interface DialogOutcome {
    /** 結果値つきの完了。 */
    class Completed(val value: Any?) : DialogOutcome

    /** 結果値を持たないキャンセル。 */
    data object Cancelled : DialogOutcome
}
